urwid/widget.py

```diff
@@ -1733,6 +1733,71 @@ def value(self):
             return 0
 
 
+class FloatEdit(Edit):
+    """Edit widget for float values"""
+
+    def valid_char(self, ch):
+        """
+        Return true for decimal digits.
+        """
+        return len(ch)==1 and ch in "0123456789."
+
+    def __init__(self,caption="",default=None):
+        """
+        caption -- caption markup
+        default -- default edit value
+        >>> FloatEdit(u"",  1.065434)
+        <FloatEdit selectable flow widget '1.065434' edit_pos=8>
+        """
+        if default is not None: val = str(default)
+        else: val = ""
+        self.__super.__init__(caption,val)
+
+    def keypress(self, size, key):
+        """
+        Handle editing keystrokes.  Remove leading zeros.
+        >>> e, size = FloatEdit(u"", 1.065434), (10,)
+        >>> e.keypress(size, 'home')
+        >>> e.keypress(size, 'delete')
+        >>> print e.edit_text
+        .065434
+        >>> e.keypress(size, 'end')
+        >>> print e.edit_text
+        .065434
+        >>> e.keypress(size, 'backspace')
+        >>> print e.edit_text
+        .06543
+        """
+        (maxcol,) = size
+        unhandled = Edit.keypress(self,(maxcol,),key)
+
+        if not unhandled:
+        # trim leading zeros
+            while self.edit_pos > 0 and self.edit_text.count('.') > 1 and \
+                  self.edit_text[-1] == ".":
+                self.set_edit_pos(self.edit_pos - 1)
+                self.set_edit_text(self.edit_text[:-1])
+
+        return unhandled
+
+    def value(self):
```

Comment:
1) I'd name it `get_value`
2) Would be cool to add an extra keyword arg `as_decimal=False` or `as_str=False` and return `Decimal` or `str` when this kwarg is `True` instead of casting it to float.

Reply:
Personally I don't like the approach of different methods to get the value.
But it is an option to do it this way, which will give the flexibility of different types and still one single method to get the value (or by property)

```python
from decimal import Decimal

class FloatEdit(object):
    def __init__(self, v, rtype=float):
        self._v = v
        self._rtype = rtype

    @property
    def v(self):
        return self._rtype(self._v)

    def get_value(self):
        return self._rtype(self._v)


fef = FloatEdit(3.14)   # floatedit float
print(fef.get_value())   # or
print(fef.v)

fed = FloatEdit(3.14, Decimal) # floatedit Decimal
print(fed.get_value())
print(fef.v)

fes = FloatEdit(3.14, str)  # floatedit str
print(fes.get_value())
print(fef.v)

```

Reply:
`rtype=float` looks *very* strange to me. It's simpler, easier to understand, and more explicit (zen of python) for the caller to convert the type, e.g. `float(fed.v)` when they need it.

Internally the only representations that make sense for `self._v` is `str` or `Decimal`. floats shouldn't be allowed to avoid the rounding UX issues. Might even want to raise an error if the user passes a float.

```python
>>> Decimal(0.01)
Decimal('0.01000000000000000020816681711721685132943093776702880859375')
```
would be very confusing to most users

Reply:
my snippet is a bit misinterpreted. Of course is the internal value a string (self.edit_text).

If it is more *Zen*  to convert by the user then I suggest:

  + only *Decimal* is returned, it is up to the user to perform other conversions from the return value if needed
  + input can be *str* or *Decimal*. Checks are performed on the input to check if it is a valid numerical value. If *float()* conversion fails it raises a ValueError
  + the *value()* raises a ValueError in case the default was None and the user did not edit the value. Currently 0 is returned (as IntEdit does) which is definitely not a correct value
  + a parameter *preserveSignificance=True* is added. If the initial value of 1234.0000 is passed and the user edits the 0-'os away to 1234. then the value will still get returned as 1234.0000

if we can agree on this then I will see to make the changes

Reply:
@wardi @hootnot +1 for me. Just keep in mind that decimal mark can be "," or "." in various locales, would be nice to tolerate that as well. Thanks!